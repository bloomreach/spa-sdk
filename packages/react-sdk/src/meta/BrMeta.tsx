/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import ReactDOM from 'react-dom';
import { MetaComment, Meta, META_POSITION_BEGIN, META_POSITION_END, isMetaComment } from '@bloomreach/spa-sdk';

interface BrMetaProps {
  meta: Meta[];
}

export class BrMeta extends React.Component<BrMetaProps> {
  private comments: Comment[] = [];
  private head?: Element | Text;
  private tailRef = React.createRef<HTMLElement>();

  componentDidMount() {
    this.head = ReactDOM.findDOMNode(this)!;
    this.renderMeta();
  }

  componentDidUpdate() {
    this.head = ReactDOM.findDOMNode(this)!;
    this.removeMeta();
    this.renderMeta();
  }

  componentWillUnmount() {
    this.removeMeta();
  }

  private removeMeta() {
    this.comments.splice(0).forEach(comment => comment.remove());
  }

  private renderMeta() {
    this.props.meta.filter(meta => isMetaComment(meta) && meta.getPosition() === META_POSITION_BEGIN)
      .forEach(this.renderMetaComment.bind(this));
    this.props.meta.filter(meta => isMetaComment(meta) && meta.getPosition() === META_POSITION_END)
      .reverse()
      .forEach(this.renderMetaComment.bind(this));
  }

  private renderMetaComment(meta: MetaComment) {
    const comment = this.head!.ownerDocument!.createComment(meta.getData());
    this.comments.push(comment);

    if (meta.getPosition() === META_POSITION_BEGIN) {
      return void this.head!.parentNode!.insertBefore(comment, this.head!);
    }

    if (!this.tailRef.current!.nextSibling) {
      return void this.tailRef.current!.parentNode!.appendChild(comment);
    }

    this.tailRef.current!.parentNode!.insertBefore(comment, this.tailRef.current!.nextSibling);
  }

  render() {
    return (
      <>
        {this.props.children}
        {this.props.meta.length ? <span style={{ display: 'none' }} ref={this.tailRef} /> : null}
      </>
    );
  }
}
