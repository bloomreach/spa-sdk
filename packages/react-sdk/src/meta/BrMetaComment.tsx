/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { MetaComment } from '@bloomreach/spa-sdk';

interface BrMetaCommentProps {
  meta: MetaComment;
}

export class BrMetaComment extends React.Component<BrMetaCommentProps> {
  private placeholder = React.createRef<HTMLElement>();
  private comment?: Comment;

  componentDidMount() {
    const placeholder = this.placeholder.current!;
    this.comment = placeholder.ownerDocument!.createComment(this.props.meta.getData());
    placeholder.parentNode!.replaceChild(this.comment, placeholder);
  }

  shouldComponentUpdate(nextProps: BrMetaCommentProps) {
    return nextProps.meta.getData() !== this.props.meta.getData();
  }

  componentDidUpdate() {
    this.comment!.replaceData(0, this.comment!.data.length, this.props.meta.getData());
  }

  componentWillUnmount() {
    this.comment!.parentNode!.replaceChild(this.placeholder.current!, this.comment!);
    delete this.comment;
  }

  render() {
    return <span style={{ display: 'none' }} ref={this.placeholder} />;
  }
}
