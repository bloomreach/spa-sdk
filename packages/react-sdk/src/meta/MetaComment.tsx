/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { MetaComment as SpaMetaComment } from '@bloomreach/spa-sdk';

interface MetaCommentProps {
  meta: SpaMetaComment;
}

export class MetaComment extends React.Component<MetaCommentProps> {
  private placeholder = React.createRef<HTMLElement>();
  private comment?: Comment;

  componentDidMount() {
    if (!this.placeholder.current) {
      return;
    }

    const placeholder = this.placeholder.current;
    this.comment = placeholder.ownerDocument!.createComment(this.props.meta.getData());
    placeholder.parentNode!.replaceChild(this.comment, placeholder);
  }

  shouldComponentUpdate(nextProps: MetaCommentProps) {
    return nextProps.meta.getData() !== this.props.meta.getData();
  }

  componentDidUpdate() {
    this.comment!.replaceData(0, this.comment!.data.length, this.props.meta.getData());
  }

  componentWillUnmount() {
    if (!this.comment) {
      return;
    }

    this.comment.remove();
  }

  render() {
    return <span style={{ display: 'none' }} ref={this.placeholder} />;
  }
}
