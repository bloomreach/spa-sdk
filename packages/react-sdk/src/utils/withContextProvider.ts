// eslint-disable-next-line import/no-extraneous-dependencies
import PropTypes from 'prop-types';
import React, { ReactElement } from 'react';

export const withContextProvider = (context?: Record<string, any>, children?: ReactElement): React.ReactElement => {
  const childContextTypes = (): Record<string, string> | undefined => context
    && Object.keys(context).reduce((obj: Record<string, any>, key: string) => {
      obj[key] = PropTypes.any;
      return obj;
    }, {});

  class ContextProvider extends React.Component {
    getChildContext(): Record<string, string> | undefined {
      return context;
    }

    render(): React.ReactNode {
      return children;
    }
  }

  (ContextProvider as any).displayName = 'ContextProvider';
  (ContextProvider as any).childContextTypes = childContextTypes();

  return React.createElement(ContextProvider, null);
};
