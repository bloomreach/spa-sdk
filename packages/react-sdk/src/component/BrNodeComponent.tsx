/*
 * Copyright 2019-2025 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Component } from '@bloomreach/spa-sdk';
import React, { useContext } from 'react';
import { BrMeta } from '../meta';
import { BrMappingContext } from './BrMappingContext';
import { BrProps } from './BrProps';

export function BrNodeComponent<T extends Component>(props: React.PropsWithChildren<BrProps<T>>): React.ReactElement {
  const context = useContext(BrMappingContext);
  const { component, children } = props;

  const mapping = component && (context[component.getName()] as React.ComponentType<BrProps>);
  const meta = component?.getMeta();

  const content = mapping ? React.createElement(mapping, props) : children;

  return React.createElement(BrMeta, { meta }, content);
}
