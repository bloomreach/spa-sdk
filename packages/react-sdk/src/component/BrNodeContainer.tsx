/*
 * Copyright 2019-2025 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useContext } from 'react';
import {
  Container,
  TYPE_CONTAINER_INLINE,
  TYPE_CONTAINER_NO_MARKUP,
  TYPE_CONTAINER_ORDERED_LIST,
  TYPE_CONTAINER_UNORDERED_LIST,
} from '@bloomreach/spa-sdk';
import { BrProps } from './BrProps';
import { BrMappingContext } from './BrMappingContext';
import { BrMeta } from '../meta';
import {
  BrContainerBox,
  BrContainerInline,
  BrContainerNoMarkup,
  BrContainerOrderedList,
  BrContainerUnorderedList,
} from '../cms';

export function BrNodeContainer(props: React.PropsWithChildren<BrProps<Container>>): React.ReactElement {
  const context = useContext(BrMappingContext);
  const { component, children } = props;

  const getMapping = (): React.ComponentType<BrProps> => {
    const type = component?.getType();

    if (type && type in context) {
      return context[type] as React.ComponentType<BrProps>;
    }

    switch (type) {
      case TYPE_CONTAINER_INLINE:
        return BrContainerInline;
      case TYPE_CONTAINER_NO_MARKUP:
        return BrContainerNoMarkup;
      case TYPE_CONTAINER_ORDERED_LIST:
        return BrContainerOrderedList;
      case TYPE_CONTAINER_UNORDERED_LIST:
        return BrContainerUnorderedList;
      default:
        return BrContainerBox;
    }
  };

  const mapping = getMapping();
  const meta = component?.getMeta();
  const content = mapping ? React.createElement(mapping, props) : children;

  return React.createElement(BrMeta, { meta }, content);
}
