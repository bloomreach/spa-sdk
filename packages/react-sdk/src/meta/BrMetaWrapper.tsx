/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { Meta, META_POSITION_BEGIN, META_POSITION_END } from '@bloomreach/spa-sdk';
import { BrMeta } from './BrMeta';

interface BrMetaWrapperProps {
  meta: Meta[];
}

function renderMeta(meta: Meta[], position: typeof META_POSITION_BEGIN | typeof META_POSITION_END) {
  return meta
    .filter(meta => position === meta.getPosition())
    .map((meta, index) => <BrMeta key={index} meta={meta} />);
}

// tslint:disable-next-line:function-name
export function BrMetaWrapper(props: React.PropsWithChildren<BrMetaWrapperProps>) {
  return (
    <>
      {renderMeta(props.meta, META_POSITION_BEGIN)}
      {props.children}
      {renderMeta(props.meta, META_POSITION_END)}
    </>
  );
}
