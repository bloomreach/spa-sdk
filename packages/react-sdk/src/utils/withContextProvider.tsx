/*
 * Copyright 2019-2025 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { ReactElement } from 'react';
import { BrPageContext } from '../page';
import { BrComponentContext, BrMappingContext } from '../component';

export const withPageContextProvider = (context?: any, children?: ReactElement): React.ReactElement => {
  if (!context) {
    return children as React.ReactElement;
  }

  return <BrPageContext.Provider value={context}>{children}</BrPageContext.Provider>;
};

export const withMappingContextProvider = (context?: any, children?: ReactElement): React.ReactElement => {
  if (!context) {
    return children as React.ReactElement;
  }

  return (
    <BrMappingContext.Provider value={context}>
      {children}
    </BrMappingContext.Provider>
  );
};

export const withComponentContextProvider = (context?: any, children?: ReactElement): React.ReactElement => {
  if (!context) {
    return children as React.ReactElement;
  }

  return <BrComponentContext.Provider value={context}>{children}</BrComponentContext.Provider>;
};
