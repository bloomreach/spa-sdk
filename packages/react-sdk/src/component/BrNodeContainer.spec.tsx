/*
 * Copyright 2019-2022 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import {
  Container,
  MetaCollection,
  Page,
  TYPE_CONTAINER_BOX,
  TYPE_CONTAINER_INLINE,
  TYPE_CONTAINER_NO_MARKUP,
  TYPE_CONTAINER_ORDERED_LIST,
  TYPE_CONTAINER_UNORDERED_LIST,
} from '@bloomreach/spa-sdk';
import { render } from '@testing-library/react';
import {
  BrContainerBox,
  BrContainerInline,
  BrContainerNoMarkup,
  BrContainerOrderedList,
  BrContainerUnorderedList,
} from '../cms';
import { BrNodeComponent } from './BrNodeComponent';
import { BrNodeContainer } from './BrNodeContainer';
import { BrMeta } from '../meta';
import { withContextProvider } from '../utils/withContextProvider';

describe('BrNodeContainer', () => {
  const props = {
    component: { getType: jest.fn(), getMeta: jest.fn() } as unknown as jest.Mocked<Container>,
    page: { isPreview: jest.fn(() => false) } as unknown as jest.Mocked<Page>,
  };

  const emptyMeta = {} as MetaCollection;

  beforeEach(() => {
    jest.resetAllMocks();

    props.component.getMeta.mockReturnValue(emptyMeta);
  });

  describe('getMapping', () => {
    beforeEach(() => {
      (BrNodeContainer as any).contextTypes = { test: () => null };
      delete (BrNodeComponent as Partial<typeof BrNodeComponent>).contextType;
    });

    it('should use container type for mapping', () => {
      render(<BrNodeContainer {...props} />);

      expect(props.component.getType).toBeCalled();
    });

    it('should render a mapped container', () => {
      props.component.getType.mockReturnValue('test' as ReturnType<Container['getType']>);
      const element = render(
        withContextProvider(
          {
            test: ({ children }: React.PropsWithChildren<typeof props>) => <div>{children}</div>,
          },
          <BrNodeContainer {...props}>
            <a />
          </BrNodeContainer>,
        ),
      );

      expect(element.asFragment()).toMatchSnapshot();
    });

    it('should render inline container', () => {
      props.component.getType.mockReturnValue(TYPE_CONTAINER_INLINE);
      const element = render(
        <BrNodeContainer {...props}>
          <a />
        </BrNodeContainer>,
      );

      const nodeMeta = render(
        <BrMeta meta={emptyMeta}>
          <BrContainerInline {...props}>
            <a />
          </BrContainerInline>
        </BrMeta>,
      );

      expect(element.container.isEqualNode(nodeMeta.container)).toBe(true);
      expect(element.asFragment()).toMatchSnapshot();
    });

    it('should render no markup container', () => {
      props.component.getType.mockReturnValue(TYPE_CONTAINER_NO_MARKUP);
      const element = render(
        <BrNodeContainer {...props}>
          <a />
        </BrNodeContainer>,
      );

      const nodeMeta = render(
        <BrMeta meta={emptyMeta}>
          <BrContainerNoMarkup {...props}>
            <a />
          </BrContainerNoMarkup>
        </BrMeta>,
      );

      expect(element.container.isEqualNode(nodeMeta.container)).toBe(true);
      expect(element.asFragment()).toMatchSnapshot();
    });

    it('should render ordered list container', () => {
      props.component.getType.mockReturnValue(TYPE_CONTAINER_ORDERED_LIST);
      const element = render(
        <BrNodeContainer {...props}>
          <a />
        </BrNodeContainer>,
      );

      const nodeMeta = render(
        <BrMeta meta={emptyMeta}>
          <BrContainerOrderedList {...props}>
            <a />
          </BrContainerOrderedList>
        </BrMeta>,
      );

      expect(element.container.isEqualNode(nodeMeta.container)).toBe(true);
      expect(element.asFragment()).toMatchSnapshot();
    });

    it('should render unordered list container', () => {
      props.component.getType.mockReturnValue(TYPE_CONTAINER_UNORDERED_LIST);
      const element = render(
        <BrNodeContainer {...props}>
          <a />
        </BrNodeContainer>,
      );

      const nodeMeta = render(
        <BrMeta meta={emptyMeta}>
          <BrContainerUnorderedList {...props}>
            <a />
          </BrContainerUnorderedList>
        </BrMeta>,
      );

      expect(element.container.isEqualNode(nodeMeta.container)).toBe(true);
      expect(element.asFragment()).toMatchSnapshot();
    });

    it('should render box container', () => {
      props.component.getType.mockReturnValue(TYPE_CONTAINER_BOX);
      const element = render(
        <BrNodeContainer {...props}>
          <a />
        </BrNodeContainer>,
      );

      const nodeMeta = render(
        <BrMeta meta={emptyMeta}>
          <BrContainerBox {...props}>
            <a />
          </BrContainerBox>
        </BrMeta>,
      );

      expect(element.container.isEqualNode(nodeMeta.container)).toBe(true);
      expect(element.asFragment()).toMatchSnapshot();
    });

    it('should render box container on an unknown type', () => {
      const element = render(<BrNodeContainer {...props} />);

      const nodeMeta = render(
        <BrMeta meta={emptyMeta}>
          <BrContainerBox {...props} />
        </BrMeta>,
      );

      expect(element.container.isEqualNode(nodeMeta.container)).toBe(true);
      expect(element.asFragment()).toMatchSnapshot();
    });
  });
});
