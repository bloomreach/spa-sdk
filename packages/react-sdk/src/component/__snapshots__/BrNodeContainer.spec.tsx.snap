// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrNodeContainer getMapping should render a mapped container 1`] = `
<DocumentFragment>
  <div>
    <a />
  </div>
</DocumentFragment>
`;

exports[`BrNodeContainer getMapping should render box container 1`] = `
<DocumentFragment>
  <div>
    <div>
      <a />
    </div>
  </div>
</DocumentFragment>
`;

exports[`BrNodeContainer getMapping should render box container on an unknown type 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;

exports[`BrNodeContainer getMapping should render inline container 1`] = `
<DocumentFragment>
  <div>
    <span>
      <a />
    </span>
  </div>
</DocumentFragment>
`;

exports[`BrNodeContainer getMapping should render no markup container 1`] = `
<DocumentFragment>
  <a />
</DocumentFragment>
`;

exports[`BrNodeContainer getMapping should render ordered list container 1`] = `
<DocumentFragment>
  <ol>
    <li>
      <a />
    </li>
  </ol>
</DocumentFragment>
`;

exports[`BrNodeContainer getMapping should render unordered list container 1`] = `
<DocumentFragment>
  <ul>
    <li>
      <a />
    </li>
  </ul>
</DocumentFragment>
`;
